primitives/sr-primitives/src/random_number_generator.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A simple random number generator that allows a stream of random numbers to be efficiently
+//! created from a single initial seed hash.
+
+use codec::{Encode, Decode};
+use crate::traits::{Hash, TrailingZeroInput};
+
+/// Random number streamer. This retains the state of the random number stream.
+///
+/// It can be saved and later reloaded using the Codec traits.
+///
+/// Example:
+/// ```
+/// use sr_primitives::traits::{Hash, BlakeTwo256};
+/// use sr_primitives::RandomNumberGenerator;
+/// let random_seed = BlakeTwo256::hash(b"Sixty-nine");
+/// let mut rng = <RandomNumberGenerator<BlakeTwo256>>::new(random_seed);
+/// assert_eq!(rng.pick_u32(100), 47);
+/// assert_eq!(rng.pick_item(&[1, 2, 3]), Some(&2));
+/// ```
+///
+/// This can use any cryptographic `Hash` function as the means of entropy-extension, and avoids
+/// needless extensions of entropy.
+#[derive(Encode, Decode)]
+pub struct RandomNumberGenerator<Hashing: Hash> {
+	current: Hashing::Output,
+	offset: u32,
+}
```

Comment:
This PRNG has an expected periodicity of ~`2^(bits(n)/2)` due to the birthday problem. For a 256-bit hash, we'd expect to reach a cycle after about 2^128 hashes - that's fine, but for a 64-bit or 80-bit hash it may be less so. I think this should be documented with a suggestion not to use short hashes.

Reply:
Even a 64-bit hash would result in a birthday only after ~2^32 bits (= 512 MB) have been drained. Given that this happens within runtime logic and generally within the same block, if you're pulling upwards of 1GB of data out of a single hash in one block, you probably have other problems to worry about than the possibility of a repeated sequence.

Reply:
I wouldn't use it like that, but I wouldn't be surprised if someone uses it in an offchain worker.

Reply:
(I put a note in there anyway)